tests/cli/test_cmds_spark_run.py

```diff
@@ -1644,3 +1655,179 @@ def test_get_aws_credentials_default_profile(mock_boto3_session, mock_use_aws_pr
+@mock.patch.object(spark_run, "get_instance_config", autospec=True)
+@mock.patch.object(spark_run, "get_aws_credentials", autospec=True)
+@mock.patch.object(spark_run, "get_docker_image", autospec=True)
+@mock.patch.object(spark_run, "get_spark_app_name", autospec=True)
+@mock.patch.object(spark_run, "auto_add_timeout_for_spark_job", autospec=True)
+@mock.patch.object(spark_run, "_parse_user_spark_args", autospec=True)
+@mock.patch(
+    "paasta_tools.cli.cmds.spark_run.spark_config.SparkConfBuilder", autospec=True
+)
+@mock.patch.object(spark_run, "configure_and_run_docker_container", autospec=True)
+@mock.patch.object(spark_run, "get_smart_paasta_instance_name", autospec=True)
+@pytest.mark.parametrize(
+    "iam_role_set,force_pod_identity,aws_creds_provided,custom_yelpsoa,should_exit_early",
+    [
+        # Service has iam_role set, no overrides
+        ("arn:aws:iam::123456789:role/test", False, None, None, 0),
+        # Service has iam_role set, you can override aws_creds_yamls
+        ("arn:aws:iam::123456789:role/test", False, True, None, 0),
+        # Force-pod-identity set, should fail as it's not in the (empty) allowed set
+        (None, "arn:aws:iam::123456789:role/test", None, None, 1),
+        # Force-pod-identity set, succeeds because test adds it to allowed set
+        (None, "arn:aws:iam::123456789:role/test", None, None, 0),
+        # Force-pod-identity set, different from set iam_role, fails as it's not in allowed set
+        (
+            "arn:aws:iam::123456789:role/test",
+            "arn:aws:iam::123456789:role/test2",
+            None,
+            None,
+            1,
+        ),
+        # Force-pod-identity set, but fails because custom soa dir provided
+        (None, "arn:aws:iam::123456789:role/test", None, "/custom/soa", 1),
+    ],
+)
+def test_paasta_spark_run_with_pod_identity(
+    mock_get_smart_paasta_instance_name,
+    mock_configure_and_run_docker_container,
+    mock_spark_conf_builder,
+    mock_parse_user_spark_args,
+    mock_auto_add_timeout_for_spark_job,
+    mock_get_spark_app_name,
+    mock_get_docker_image,
+    mock_get_aws_credentials,
+    mock_get_instance_config,
+    mock_load_system_paasta_config_spark_run,
+    mock_load_system_paasta_config_utils,
+    mock_validate_work_dir,
+    mock_get_all_iam_roles_for_service,
+    iam_role_set,
+    force_pod_identity,
+    aws_creds_provided,
+    custom_yelpsoa,
+    should_exit_early,
+):
+    args = argparse.Namespace(
+        work_dir="/tmp/local",
+        cmd="USER=test spark-submit test.py",
+        build=True,
+        image=None,
+        enable_compact_bin_packing=False,
+        disable_compact_bin_packing=False,
+        service="test-service",
+        instance="test-instance",
+        cluster="test-cluster",
+        pool="test-pool",
+        yelpsoa_config_root=custom_yelpsoa or spark_run.DEFAULT_SOA_DIR,
+        aws_credentials_yaml="/path/to/creds" if aws_creds_provided else None,
```

Comment:
i usually try to avoid branches in tests to keep them simple, but this (and the others) are probably fine

Reply:
It's a little funky but I wanted to avoid copying the entire test several times. 🤷